src/game/client/shadow/w3dvolumetricshadow.cpp

```diff
@@ -0,0 +1,503 @@
+                                index1 = -1;
+                            }
+                        }
+                    }
+                }
+
+                if (index1 != -1 && index2 != -1) {
+                    // found a shared edge
+                    for (int edge = 0; edge < 3; ++edge) {
+                        if (m_polyNeighbors[i].neighbor[edge].neighborIndex == -1) {
+                            m_polyNeighbors[i].neighbor[edge].neighborIndex = (short)j;
+                            m_polyNeighbors[i].neighbor[edge].neighborEdgeIndex[0] = (short)index1;
+                            m_polyNeighbors[i].neighbor[edge].neighborEdgeIndex[1] = (short)index2;
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+bool W3DShadowGeometryMesh::Allocate_Neighbors(int num_polys)
+{
+    captainslog_assert(m_numPolyNeighbors == 0);
+    captainslog_assert(m_polyNeighbors == nullptr);
+
+    m_polyNeighbors = new PolyNeighbor[num_polys];
+
+    if (!m_polyNeighbors) {
+        captainslog_assert(0);
+        return false;
+    }
+
+    m_numPolyNeighbors = num_polys;
+    return true;
+}
+
+void W3DShadowGeometryMesh::Delete_Neighbors()
+{
+
+    if (m_polyNeighbors) {
+        delete[] m_polyNeighbors;
+        m_polyNeighbors = nullptr;
+        m_numPolyNeighbors = 0;
+    }
+
+    captainslog_assert(m_numPolyNeighbors == 0);
+    captainslog_assert(m_polyNeighbors == nullptr);
+}
+
+Vector3 *W3DShadowGeometryMesh::Get_Polygon_Normal(int index) const
+{
+    captainslog_assert(m_polygonNormals);
+
+    return &m_polygonNormals[index];
+}
+
+void W3DShadowGeometryMesh::Build_Polygon_Normals()
+{
+    if (m_polygonNormals == nullptr) {
+        Vector3 *normals = new Vector3[m_numPolygons];
+
+        for (int i = 0; i < m_numPolygons; ++i) {
+            Get_Polygon_Normal(i, &normals[i]);
+        }
+
+        m_polygonNormals = normals;
+    }
+}
+
+void W3DShadowGeometryMesh::Get_Polygon_Normal(int index, Vector3 *normal) const
+{
+
+    if (m_polygonNormals) {
+        *normal = m_polygonNormals[index];
+    }
+
+    short index_list[3];
+    Get_Polygon_Index(index, index_list);
+
+    Vector3 *v0 = Get_Vertex(index_list[0]);
+    Vector3 *v1 = Get_Vertex(index_list[1]);
+    Vector3 *v2 = Get_Vertex(index_list[2]);
+
+    Vector3 b = *v1 - *v0;
+    Vector3 a = *v1 - *v2;
+
+    Vector3::Normalized_Cross_Product(a, b, normal);
+}
+
+void W3DShadowGeometryMesh::Get_Polygon_Index(int polygon_index, short *index_list) const
+{
+    const TriIndex *v = &m_polygons[polygon_index];
+    index_list[0] = m_parentVerts[v->I];
+    index_list[1] = m_parentVerts[v->J];
+    index_list[2] = m_parentVerts[v->K];
+}
+
+W3DShadowGeometry::W3DShadowGeometry() : m_name{}, m_meshCount(0), m_numTotalsVerts(0) {}
+
+void W3DShadowGeometry::Set_Name(const char *name)
+{
+    memset(m_name, 0, sizeof(m_name));
+
+    strncpy(m_name, name, sizeof(m_name) - 1);
```

Comment:
Could use `strlcpy` here, unless the string really needs zero-ed out after last character.